Feature/multiple encodings handled

jeremyelalouf opened the pull request:
# Hi everyone 👋🏻

Here, you will find the integration of this [Gist](https://gist.github.com/rojas-diego/04d9c4e3fff5f8374f29b9b738d541ef) made by @rojas-diego, which adds support for multiple encodings.

## Actually

The only supported encoding was UTF-8.
In cases where the editor sends updates encoded in UTF-16,
the mirror of the user's workspace goes out of sync, leading to a server crash.

Furthermore, UTF-16 is the default and mandatory encoding for the protocol,
and servers must support it.

Therefore, it is imperative to ensure its support.

## Also

To avoid any unnecessary conversion, the server negotiates the encoding
with the client during the initialization phase.
This allows the server to choose its preferred encoding in cases where it is available.

See: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocuments

jeremyelalouf commented:
By the way, do you think that you could run the CI ? It would be great for me to assure that everything works fine.

Thanks @McPatate ! 😄

McPatate commented:
Hey @jeremyelalouf thanks for the PR!

> run the CI ?

I haven't been able to find a way to circumvent the fact that secrets are not injected in workflows coming from forks. If you could run it manually it'd be nice, otherwise I'll see later myself if this causes issues.

Did you test your code manually & can confirm it works as intended?

McPatate commented on crates/llm-ls/src/main.rs:
```diff
@@ -135,6 +135,7 @@ struct LlmService {
     workspace_folders: Arc<RwLock<Option<Vec<WorkspaceFolder>>>>,
     tokenizer_map: Arc<RwLock<HashMap<String, Arc<Tokenizer>>>>,
     unauthenticated_warn_at: Arc<RwLock<Instant>>,
+    position_encoding: Arc<RwLock<PositionEncodingKind>>,
```
You could also implement an `to_lsp_type` method on your `document::PositionEncodingKind`

McPatate commented on crates/llm-ls/src/document.rs:
```diff
@@ -1,9 +1,8 @@
-use ropey::Rope;
-use tower_lsp::lsp_types::Range;
+use ropey::{Error as RopeyError, Rope, RopeSlice};
```
```suggestion
use ropey::{Rope, RopeSlice};
```

McPatate commented on crates/llm-ls/src/document.rs:
```diff
@@ -1,9 +1,8 @@
 
-use crate::error::Result;
-use crate::get_position_idx;
+use crate::error::{Error as LspError, Result};
```
```suggestion
use crate::error::{Error, Result};
```

McPatate commented on crates/llm-ls/src/document.rs:
```diff
@@ -148,73 +158,422 @@ impl Document {
+                    self.tree = Some(self
+                                .parser
+                                .parse(self.text.to_string(), Some(tree))
+                                .expect("parse should always return a tree when the language was set and no timeout was specified"));
```
```suggestion
                                .expect("parse should always return a tree when the language was set and no timeout was specified"));
```
This should be handled properly

jeremyelalouf replied:
I did not understand what you meant by this, how do you want me to handle this case ?

rojas-diego replied:
I don't think this should cause any issues. The tree is `None` if no language was set and we don't set a timeout. I believe these are the only two ways the `parse` method can panic. In any case @jeremyelalouf, you can always propagate the error upwards.

jeremyelalouf replied:
I do agree that the use of any `expect`s, `unwrap`s, or `panic!` should be minimized, as mentioned. 

> It's probably better to propagate an error message to the client in case we add a timeout in the future.

As you suggested, Diego, if a potential timeout in the future was set, having a predefined custom error type, such as `Error::TreeSitterParseError`, is advantageous.

This would provide additional details for the user.

jeremyelalouf commented:
Hi @McPatate,

## Concerning the CI

Firstly, I apologize for not being aware of the issues you've experienced with Actions and forks.
As you can see, as a result I've initiated the CI on [my fork](https://github.com/jeremyelalouf/llm-ls/actions/runs/8170436223), and I'm awaiting a runner to pick up my jobs.

I would also add that I didn't know that I had access to self hosted runners of an Organization by forking their repository, it could be dangerous don't you think ?

## About my tests

I've taken measures to ensure that I haven't introduced any issues and that everything is functioning as expected.

To achieve this, I added tests for the document::Document struct in the document.rs file.
Additionally, I conducted manual functional tests in an editor to compare the behavior of the current version with my fixed version.

Here's a video demonstrating the previous behavior of the LSP when handling non-supported encodings, leading to a crash.

https://github.com/huggingface/llm-ls/assets/75589941/63e4a244-1d15-4a56-abc6-932913177c9d


Then, another video showcases the current behavior of the LSP after integrating the fixed code. As you can see, there are no more crashes to report 🫡.


https://github.com/huggingface/llm-ls/assets/75589941/2843b197-cbe1-4d5c-9d4e-db4f914f6936

Unfortunately, I didn't have the time to run testbed locally, which is why I requested the CI to assist me with this aspect.

---

I kindly request your understanding that, despite the checks I've performed, I cannot fully guarantee the stability of the current version.
Especially in a production environment where the codebase might be extensive, and the LSP particularly busy.

Thanks again 😄 !

McPatate reviewed (commented):
This looks very nice, thanks a lot for the hard work!

Regarding type conversion from the custom type to the `lsp_types`, here is a small example of what I mean:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=91e9f5a118e2616dee680e1fdafd05dc

I've left comments on what I think needs some work, but lgtm other than that!

McPatate commented:
> I would also add that I didn't know that I had access to self hosted runners of an Organization by forking their repository, it could be dangerous don't you think ?

You don't that's & that's what's difficult when having external contributions 😄 

Appreciate you reproducing the bug & displaying the fix!

> I kindly request your understanding that, despite the checks I've performed, I cannot fully guarantee the stability of the current version.
> Especially in a production environment where the codebase might be extensive, and the LSP particularly busy.

No problem, this looks good enough to me. I'd appreciate if you can run testbed, but not a hard requirement.

jeremyelalouf commented:
> You don't that's & that's what's difficult when having external contributions 😄

Totally makes sense 😂 !

I will then run try to run `testbed` locally once I've fixed all your comments.

Cheers !

jeremyelalouf commented on crates/llm-ls/src/main.rs:
```diff
@@ -532,6 +533,30 @@ impl LlmService {
 impl LanguageServer for LlmService {
     async fn initialize(&self, params: InitializeParams) -> LspResult<InitializeResult> {
         *self.workspace_folders.write().await = params.workspace_folders;
+        let position_encoding = params
+            .capabilities
+            .general
+            .and_then(|general_capabilities| {
+                general_capabilities
+                    .position_encodings
+                    .map(|encodings| {
+                        if encodings.contains(&PositionEncodingKind::UTF8) {
+                            PositionEncodingKind::UTF8
+                        } else if encodings.contains(&PositionEncodingKind::UTF16) {
+                            PositionEncodingKind::UTF16
+                        } else if encodings.contains(&PositionEncodingKind::UTF32) {
+                            PositionEncodingKind::UTF32
+                        } else {
+                            // Because UTF-16 is the only mandatory encoding that the client must support,
+                            // we will use it as the default encoding.
+                            // See: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocuments
+                            PositionEncodingKind::UTF16
+                        }
+                    })
+            });
+
+        *self.position_encoding.write().await = position_encoding.as_ref().unwrap().clone();
```
Actually, here `general_capabilities.position_encodings` is an array containing the different encodings supported by the client.

That's why it was interesting to choose the more convenient one to avoid unneeded conversions in the `apply_content_change` function.

jeremyelalouf commented:
Hello again @McPatate,

I've just pushed the fix of almost all comments you've made.
The only unanswered questions I had were in response to your comments, [here](https://github.com/huggingface/llm-ls/pull/88#discussion_r1515178526) and [here](https://github.com/huggingface/llm-ls/pull/88#discussion_r1515251293).

Keep in mind that for the first one, I did, however, offer a solution that I felt was appropriate in my last commit.

Have a good night !

McPatate commented on crates/llm-ls/src/error.rs:
```diff
@@ -35,10 +35,12 @@ pub enum Error {
+    OutOfBoundLine(usize, usize),
     #[error("slice out of bounds: {0}..{1}")]
     OutOfBoundSlice(usize, usize),
+    #[error("range out of bounds: {0:?}")]
+    InvalidRange(Range),
```
Everything is sorted by alphabetical order, can you move the error to its right place please? 😇

jeremyelalouf replied:
Certainly! Apologies for any oversights and errors; I was rushing a bit too much...

McPatate replied:
No worries, that's the point of reviews 🤗

McPatate commented on crates/llm-ls/src/error.rs:
```diff
@@ -35,10 +35,12 @@ pub enum Error {
     OutOfBoundIndexing(usize),
-    #[error("line out of bounds: {0}")]
-    OutOfBoundLine(usize),
+    #[error("line out of bounds: {0}..{1}")]
```
```suggestion
    #[error("line out of bounds: {0} >= {1}")]
```
No ?

McPatate commented on crates/llm-ls/src/error.rs:
```diff
@@ -53,6 +55,10 @@ pub enum Error {
     UnknownBackend(String),
+    #[error("unknown encoding kind: {0}")]
+    UnknownEncodingKind(String),
+    #[error("no encoding kind provided by the client")]
+    EncodingKindMissing
```
Same here!

McPatate commented on crates/llm-ls/src/main.rs:
```diff
@@ -541,6 +553,7 @@ impl LanguageServer for LlmService {
                 text_document_sync: Some(TextDocumentSyncCapability::Kind(
                     TextDocumentSyncKind::INCREMENTAL,
                 )),
+                position_encoding: Some(position_encoding.to_lsp_types()),
```
```suggestion
                position_encoding: Some(position_encoding.to_lsp_type()),
```

McPatate reviewed (commented):
looking good!

jeremyelalouf commented:
There you go, @McPatate! I believe I've addressed all your comments.

Please let me know if I've overlooked anything or if you have any additional feedback !

McPatate reviewed (approved):
lgtm! Thank you for the hard work :)

jeremyelalouf commented:
With great pleasure !

It was a great experience to contribute to this project 🥳

(and thanks to my best Astek @rojas-diego)
